kong/observability/logs.lua

```diff
@@ -0,0 +1,179 @@
+      "[observability] OpenTelemetry logs failed reading dynamic log level. " ..
+      "Using log level: " .. level_str .. " from configuration."
+    )
+    level = constants_log_levels[level_str]
+  end
+
+  return level
+end
+
+
+-- needed because table.concat doesn't like booleans
+local function concat_tostring(tab)
+  local tab_len = #tab
+  if tab_len == 0 then
+    return ""
+  end
+
+  for i = 1, tab_len do
+    local value = tab[i]
+
+    if value == ngx_null then
+      value = "nil"
+    else
+      value = tostring(value)
+    end
+
+    logline_buf:put(value)
+  end
+
+  return logline_buf:get()
+end
+
+
+local function generate_log_entry(request_scoped, log_level, log_str, request_id, debug_info)
+
+  local span_id
+
+  if request_scoped then
+    -- add tracing information if tracing is enabled
+    local active_span = kong and kong.tracing and kong.tracing.active_span()
+    if active_span then
+      span_id = active_span.span_id
+    end
+  end
+
+  local attributes = {
+    ["request.id"] = request_id,
+    ["introspection.current.line"] = debug_info.currentline,
+    ["introspection.name"] = debug_info.name,
+    ["introspection.namewhat"] = debug_info.namewhat,
+    ["introspection.source"] = debug_info.source,
+    ["introspection.what"] = debug_info.what,
+  }
+
+  local now_ns = time_ns()
+  return {
+    time_unix_nano = now_ns,
+    observed_time_unix_nano = now_ns,
+    log_level = log_level,
+    body = log_str,
+    attributes = attributes,
+    span_id = span_id,
+  }
+end
+
+
+local function get_request_log_buffer()
+  local log_buffer = ngx.ctx[NGX_CTX_REQUEST_LOGS_KEY]
+  if not log_buffer then
+    log_buffer = table_new(10, 0)
+    ngx.ctx[NGX_CTX_REQUEST_LOGS_KEY] = log_buffer
+  end
+  return log_buffer
+end
+
+
+function _M.maybe_push(stack_level, log_level, ...)
+  -- WARNING: do not yield in this function, as it is called from ngx.log
+
+  -- Early return cases:
+
+  -- log level too low
+  if configured_log_level() < log_level then
+    return
+  end
+
+  local log_buffer, max_logs
+  local request_id = request_id_get()
+  local request_scoped = request_id ~= nil
+
+  -- get the appropriate log buffer depending on the current context
+  if request_scoped then
+    log_buffer = get_request_log_buffer()
+    max_logs = MAX_REQUEST_LOGS
+
+  else
+    log_buffer = worker_logs
+    max_logs = MAX_WORKER_LOGS
+  end
+
+  -- return if log buffer is full
+  if #log_buffer >= max_logs then
+    return
+  end
+
+  -- no (or empty) log line
+  local args = table_pack(...)
+  local log_str = concat_tostring(args)
+  if log_str == "" then
+    return
+  end
+
+  -- generate & push log entry
+  local debug_info = debug.getinfo(stack_level, "nSl")
+  local log_entry = generate_log_entry(request_scoped, log_level, log_str, request_id, debug_info)
+  table.insert(log_buffer, log_entry)
+end
+
+
+function _M.get_worker_logs()
+  local wl = deep_copy(worker_logs)
+
+  table_clear(worker_logs)
+  return wl
+end
```

Comment:
Just a quick question, `deep_copy and clearing table` is for better performance?

Reply:
`deep_copy` is because we pass the table to a queue for processing asynchronously, so we need to make a copy, `clear` is because we consume the worker-scoped logs when they're read, this is not needed for request-scoped logs because they're kept in `ngx.ctx`. Did I answer your question?

Reply:
How about this?

```lua
local wl = worker_logs
worker_logs = table_new(CAPACITY, 0)
return wl
```

Reply:
done